### PR TITLE
Enable SSL verification by default

### PR DESCRIPTION
**Pull request by jes**
### Summary
Enable SSL verification by default by not setting SSL_verify_mode to 0x00. Add an option to Mojo::UserAgent ("verify") to allow people to explicitly choose non-verifying behaviour. Make Test::Mojo automatically turn off verification so SSL testing still works as before.

I don't expect this is enough to merge straight away, but I think it's a good starting point.

### Motivation
Prevent people from inadvertently writing insecure software.

### References
Issue #1223 "Mojolicious does not check SSL certificates by default".

### CODE REVIEW

File: lib/Mojo/UserAgent.pm [hunk 1]
```diff
@@ -16,6 +16,7 @@ use constant DEBUG => $ENV{MOJO_USERAGENT_DEBUG} || 0;
 
 has ca              => sub { $ENV{MOJO_CA_FILE} };
 has cert            => sub { $ENV{MOJO_CERT_FILE} };
+has verify          => sub { $ENV{MOJO_VERIFY} };
```
Comment (kraih) [inline lib/Mojo/UserAgent.pm]: I think the option should simply be called `insecure`, like in curl.

File: lib/Mojo/UserAgent.pm [hunk 2]
```diff
@@ -1025,6 +1025,13 @@ L<Mojo::Promise> object instead of accepting a callback.
     warn "Connection error: $err";
   })->wait;
 
+=head2 verify
+
+  my $verify_mode = $ua->verify;
+  $ua->verify(0); # disable TLS certificate verification
+
+Set the TLS verify_mode. See SSL_verify_mode in IO::Socket::SSL.
```
Comment (kraih) [inline lib/Mojo/UserAgent.pm]: The example and description are inconsistent with other attribute descriptions.

File: lib/Test/Mojo.pm [hunk 3]
```diff
@@ -17,7 +17,7 @@ use Mojo::Util qw(decode encode);
 use Test::More ();
 
 has [qw(message success tx)];
-has ua => sub { Mojo::UserAgent->new->ioloop(Mojo::IOLoop->singleton) };
+has ua => sub { Mojo::UserAgent->new(verify => 0)->ioloop(Mojo::IOLoop->singleton) };
```
Comment (kraih) [inline lib/Test/Mojo.pm]: Looks like you did not perltidy anything.

File: t/mojo/certs/selfsigned.crt [hunk 4]
```diff
@@ -0,0 +1,32 @@
+-----BEGIN CERTIFICATE-----
```
Comment (kraih) [inline t/mojo/certs/selfsigned.crt]: There should be instructions in `t/mojo/ioloop_tls.t` for how to regenerate all certs. But are you sure the existing certs don't already cover this case?

File: t/mojo/user_agent_tls.t [hunk 5]
```diff
@@ -99,4 +100,22 @@ is $ua->ioloop->stream($tx->connection)->handle->get_cipher, 'AES256-SHA',
...
+$daemon = Mojo::Server::Daemon->new(
+  app    => app,
+  ioloop => Mojo::IOLoop->singleton,
+  silent => 1
+);
+$listen
+  = 'https://127.0.0.1'
+  . '?cert=t/mojo/certs/selfsigned.crt'
+  . '&key=t/mojo/certs/selfsigned.key';
+$port = $daemon->listen([$listen])->start->ports->[0];
```
Comment (kraih) [inline t/mojo/user_agent_tls.t]: Is this really a new test case? Or is it just doing the same as one of the others?

**Comment by kraih** [comment]
You forgot `Mojolicious::Command::get` and `ojo`.

**Comment by kraih** [comment]
How is the CA situation after this change? On which operating systems are system certs used by default? Where do we need to configure and use something like `Mozilla::CA`? How should we document the situation?

File: lib/Mojo/UserAgent.pm [hunk 6]
```diff
@@ -503,8 +504,7 @@ L<Mojo::UserAgent> implements the following attributes.
   $ua    = $ua->ca('/etc/tls/ca.crt');
 
 Path to TLS certificate authority file used to verify the peer certificate,
-defaults to the value of the C<MOJO_CA_FILE> environment variable. Also
-activates hostname verification.
+defaults to the value of the C<MOJO_CA_FILE> environment variable.
```
Comment (kraih) [inline lib/Mojo/UserAgent.pm]: There are many more places that need this change.

File: lib/Mojo/UserAgent.pm (hunk 1 above)
Comment (kraih) [reply lib/Mojo/UserAgent.pm]: And be a boolean attribute of course. With `-k/--insecure` flag in `Mojolicious::Command::get`.

Comment (kraih) [reply lib/Mojo/UserAgent.pm]: And it appears the actual attribute accessor does not have a test yet.

**Comment by kraih** [comment]
We usually require a deprecation period for breaking changes, and this is going to break some applications. But this change does fall under the security exception in our rules, so we are a little more flexible.

**Comment by kraih** [comment]
Btw. I do welcome competing pull requests for this. If anyone else has more time to finish a polished proposal, please do. Right now this issue has momentum in the community, we shouldn't risk letting it go stale and forgotten.

**Comment by kraih** [comment]
And for the record, after this patch has been applied on openSUSE Tumbleweed and macOS all requests to HTTPS sites with **valid** certificates do no longer work by default.
```
$ mojo get https://mojolicious.org
SSL connect attempt failed error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed
 at lib/Mojolicious/Command/get.pm line 76.
```

**Comment by jes** [comment]
Many thanks for fixing this!

I wasn't ignoring your feedback, I was going to get to it next time I worked on this. Glad it's all sorted now :)

**Pull request closed without merging**